File: Data Structures & Algorithms/invert-a-binary-tree/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public TreeNode invertTree(TreeNode root) {
        //case 1: no children -> do nothing
        //case 2: 1 child -> switch pointers
        //case 3: 2 children -> switch pointers

        //invertTree(root.left);
        //root.left = root.right;
        //root.right = oldLeft;
    if(root == null) {
        return null;
    }
        if(root.left != null) {
            invertTree(root.left);
            if(root.right != null) {
                invertTree(root.right);
            }
            if(root.left != null && root.right != null) {
                TreeNode oldRight = root.right;
                root.right = root.left;
                root.left = oldRight;
            }
            else if(root.left == null && root.right != null) {
                root.left = root.right;
                root.right = null;
            }
            else if(root.left != null && root.right == null) {
                root.right = root.left;
                root.left = null;
            }
            
        }
        else if(root.right != null) {
            invertTree(root.right);
            if(root.left != null) {
                invertTree(root.left);
            }
            TreeNode oldLeft = root.left;
            root.left = root.right;
            root.right = oldLeft;
        }
        else if(root.left == null && root.right == null) {
            return root;
        }
        return root;

    }
}

File: Data Structures & Algorithms/balanced-binary-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    left = 0
    right = 0
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        if not root:
            return True
        left = self.getHeight(root.left)
        right = self.getHeight(root.right)
        if abs(left - right) > 1:
            return False
        else: 
            return self.isBalanced(root.left) and self.isBalanced(root.right)

    def getHeight(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        return 1 + max(self.getHeight(root.left), self.getHeight(root.right))